### PR TITLE
Fix a scheduler e2e bug on Preemption

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/kind bug
/kind failing-test
/sig scheduling

**What this PR does / why we need it**:

Two parts in this PR:

- The first one removes usage of assigning `.spec.nodeName` directly; instead, use NodeAffinity to let it go through scheduler. BTW: a lot of "pod ran to completion" errors surfaced recently, I believe @msau42 and @liggitt encountered this as well.
- The second part resolves an "index out of range" problem

**Which issue(s) this PR fixes**:

Part of #88441.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/scheduling/preemption.go [hunk 1]
```diff
@@ -214,14 +214,14 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
 			if len(pods) == 0 {
 				priorityName = lowPriorityClassName
 			}
-			pods[i] = createPausePod(f, pausePodConfig{
+			pods = append(pods, createPausePod(f, pausePodConfig{
```
Comment (alculquicondor) [inline test/e2e/scheduling/preemption.go]: How come this was passing before?

File: test/e2e/scheduling/predicates.go [hunk 2]
```diff
@@ -836,11 +836,13 @@ func initPausePod(f *framework.Framework, conf pausePodConfig) *v1.Pod {
 				},
 			},
 			Tolerations:                   conf.Tolerations,
-			NodeName:                      conf.NodeName,
 			PriorityClassName:             conf.PriorityClassName,
 			TerminationGracePeriodSeconds: &gracePeriod,
 		},
 	}
+	if len(conf.NodeName) != 0 {
```
Comment (alculquicondor) [inline test/e2e/scheduling/predicates.go]: why do we want this to go through the scheduler?

File: test/e2e/scheduling/preemption.go (hunk 1 above)
Comment (Huang-Wei) [reply test/e2e/scheduling/preemption.go]: No idea. I'm curious whether it really ran before - which looks like an obvious bug.

Comment (alculquicondor) [reply test/e2e/scheduling/preemption.go]: This is an obvious fix, but I'm not so sure about the other one. The scheduler should work if some pods are directly assigned. Unless this is related to #88389 (the fact that we have different event handlers).

Comment (Huang-Wei) [reply test/e2e/scheduling/preemption.go]: Yup, I don't have clear clues why the other fix _can_ resolve the problem in my env. Let me dig more.

File: test/e2e/scheduling/predicates.go (hunk 2 above)
Comment (Huang-Wei) [reply test/e2e/scheduling/predicates.go]: To not block the main release, this acts more as a workaround. So I'm extracting this section into a separate commit and leave a TODO so as to buy us some time to find out the real root cause in the future.

**Comment by Huang-Wei** [comment]
It seems /lgtm doens't work 😕

**Comment by droslean** [comment]
We need to cherry-pick this PR to 1.18 branch.

**Pull request merged**